### PR TITLE
GlobalCollect: Enable Google Pay and Apple Pay payment methods

### PR DESCRIPTION
**Pull request by gasb150**
Summary:
---------------------------------------
In order to add  Apple Pay and Google Pay support to GlobalCollect
this commit enables the option to add those payment methods for use in the add_payment method.

Local Tests:
---------------------------------------
41 tests, 215 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed

Remote Tests:
--------------------------------------
37 tests, 93 assertions, 1 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
97.2973% passed

RuboCop:
---------------------------------------
728 files inspected, no offenses detected

### CODE REVIEW

File: test/remote/gateways/remote_global_collect_test.rb
Comment (javierpedrozaing) [inline test/remote/gateways/remote_global_collect_test.rb:15]: :question:
you're using the number '4567350000427977' for google_pay and apple_pay, is it by a special reason o  we can use also some of these [test data](https://docs.spreedly.com/reference/test-data/ ) ?

**Review by javierpedrozaing — COMMENTED** [review]
Great implementation :muscle:, just I leave some comments for your consideration.

File: test/remote/gateways/remote_global_collect_test.rb
Comment (gasb150) [reply test/remote/gateways/remote_global_collect_test.rb:15]: Yes, sure, you could use the test card of [test data](https://docs.spreedly.com/reference/test-data/), the way to change between successful or unsuccessful transactions with apple/google pay for GlobalCollect is in the amount value.

File: lib/active_merchant/billing/gateways/global_collect.rb
Comment (gasb150) [reply lib/active_merchant/billing/gateways/global_collect.rb:284]: Yes, let's try your suggestion.

**Comment by curiousepic** [comment]
@gasb150 is there a reason this was put back to Draft status? Is it still ready for final review?

**Comment by gasb150** [comment]
> @gasb150 is there a reason this was put back to Draft status? Is it still ready for final review?

Sorry, David, this PR was moved to Draft to properly made the changes needed to work with Core.

File: lib/active_merchant/billing/gateways/global_collect.rb
Comment (Heavyblade) [inline lib/active_merchant/billing/gateways/global_collect.rb:273]: 💬  Comment:

It's just matter of personal taste but I think that just one assignation is more clear or "Rubyist", something like:

``` Ruby
post['cardPaymentMethodSpecificInput'] = specifics_inputs.merge({
'card' => {  'cvv' => payment.verification_value,
             'cardNumber' => payment.number,
             'expiryDate' => expirydate,
             'cardholderName' => payment.name
           }
})
```

File: lib/active_merchant/billing/gateways/global_collect.rb
Comment (Heavyblade) [inline lib/active_merchant/billing/gateways/global_collect.rb:285]: ❓  Are you checking payment_cryptogram method and value to check if the payment method is a NetworkTokenizationCreditCard  ?

File: lib/active_merchant/billing/gateways/global_collect.rb
Comment (Heavyblade) [inline lib/active_merchant/billing/gateways/global_collect.rb:284]: ❓ Question:

I'm trying to think of a possibility when data is not going to be assigned that justifies this default value, do you know a case?

File: test/remote/gateways/remote_global_collect_test.rb
Comment (Heavyblade) [inline test/remote/gateways/remote_global_collect_test.rb:32]: ❓  Question:

Maybe I'm wrong, but pan_only google pay tokens are going to be passed as instances of CreditCard not NetworkTokenization ?

File: test/remote/gateways/remote_global_collect_test.rb
Comment (Heavyblade) [inline test/remote/gateways/remote_global_collect_test.rb:100]: ❓  Question:

Along with the comment above related to @google_pay_pan_only should be a CreditCard, shouldn't the customer param be set as "google_customer_id" instead "customer".

File: test/unit/gateways/global_collect_test.rb
Comment (Heavyblade) [inline test/unit/gateways/global_collect_test.rb:93]: ❓  Why that if ?? => if the test fails because there is not mobilePaymentMethodSpecificInput that's part of test check.

File: test/unit/gateways/global_collect_test.rb
Comment (Heavyblade) [inline test/unit/gateways/global_collect_test.rb:97]: 💬  Comment:

The is no value in the testing purchase and authorize because purchase basically makes a delegate request to authorize and doesn't modify any params so at the end you are testing the same code twice.

**Review by Heavyblade — COMMENTED** [review]
@gasb150 we are very close 👍 , leaved some comments for your cosideration

File: test/unit/gateways/global_collect_test.rb
Comment (gasb150) [reply test/unit/gateways/global_collect_test.rb:93]: This is to check if the value is present and validate if the test was made with google_pay because '320' is the value of Google for paymentProductId.

File: test/unit/gateways/global_collect_test.rb
Comment (gasb150) [reply test/unit/gateways/global_collect_test.rb:97]: You are right. Test purchase tests an authorize and tests a capture.

File: test/unit/gateways/global_collect_test.rb
Comment (gasb150) [reply test/unit/gateways/global_collect_test.rb]: Actually I'm not seeing a reason to keep that if, I'll remove it.

**Review by javierpedrozaing — COMMENTED** [review]
LGTM

File: test/remote/gateways/remote_global_collect_test.rb
Comment (gasb150) [reply test/remote/gateways/remote_global_collect_test.rb]: Yes, you are right.

File: lib/active_merchant/billing/gateways/global_collect.rb
Comment (gasb150) [reply lib/active_merchant/billing/gateways/global_collect.rb]: With, considering what I'm testing, I think I need to test the cases you mention.

**Comment by curiousepic** [comment]
I'm getting failures for both remote success test cases - discussing details in Slack.

Also, FYI there was one Rubocop offense - be sure to run it again after any changes during review.

**Review by curiousepic — APPROVED** [review]
Nice, that did it. Looks good, will merge soon.

**Comment by curiousepic** [comment]
Approved a little too quickly after the fix, there are a couple concerns I'll discuss elsewhere before merging this.

File: lib/active_merchant/billing/credit_card.rb
Comment (curiousepic) [inline lib/active_merchant/billing/credit_card.rb:141]: I'm not sure if this is used in a technical way, but best to keep the pattern:
```suggestion
      # @return [Boolean]
```

File: test/unit/gateways/global_collect_test.rb
Comment (Heavyblade) [inline test/unit/gateways/global_collect_test.rb:117]: 💬  Comment:

Given this PR has a lot of logic on the `add_decrypted_payment_data` and `add_order` methods, could be a good idea to test here different scenarios to ensure that "mobilePaymentMethodSpecificInput" or "cardPaymentMethodSpecificInput" gets populated properly

**Review by Heavyblade — COMMENTED** [review]
@gasb150 looks pretty good, left just one comment that I consider could be an improvement

File: test/unit/gateways/global_collect_test.rb
Comment (gasb150) [reply test/unit/gateways/global_collect_test.rb:246]: Ok, then I'll add more tests related to those methods.

**Pull request closed without merging**
